csrc/runtime/fusion_kernel_runtime.cpp

```diff
@@ -330,6 +330,45 @@ KernelArgumentHolder FusionKernelRuntime::runWithInputs(
   return fusion_outputs;
 }
 
+std::vector<KernelArgumentHolder> FusionKernelRuntime::prepareInputs(
+    const KernelArgumentHolder& args) const {
+  std::vector<KernelArgumentHolder> all_runtime_inputs;
+  const int64_t num_groups = numGroups();
+  all_runtime_inputs.reserve(num_groups);
+
+  ArgumentManager args_manager(
+      args, runtime_workspace_, segmented_fusion_->inputs());
+
+  // group should share cache id.
+  const auto group_cache_id = args.getCacheId();
+
+  for (int64_t run_order_id = 0; run_order_id < num_groups; ++run_order_id) {
+    auto group_to_run = runtime_workspace_.group_run_order.at(run_order_id);
+    // TODO: index mode should be updated per segmented kernel
+    // Prepare input vector
+    all_runtime_inputs.push_back(
```

Comment:
Alternatively, we can index all_runtime_inputs by group_id instead of run_order_id. This way, when you compileKernel these segments, you don't need to read run_order at all.

Reply:
That seems to make the code simpler. Thanks for the suggestion.

Reply:
On second thought, it doesn't seem to buy much since we need `run_order` anyway as long as `compileFusionParallel` compiles segments in the run order. That doesn't seem to be required, but I don't see any particular motivation to change that.

Reply:
> compileFusionParallel compiles segments in the run order

Just to make sure we are on the same page, there's shape inference and there's compilation. Shape inference needs to be in the run order. Compilation doesn't and in fact runs in parallel. Therefore, the sage of run_order at https://github.com/NVIDIA/Fuser/pull/4373/files#diff-67ea17ca8b2cc538ed907fe9c9546fe45fe300237d1302379a14a0c0c0e79fcbR423 is unnecessary. 

I can clean this up after this PR.

Reply:
Right, it doesn't need to be the same order for functionality, but I just feel like it makes to sense to keep the order for debugging. What I commonly do is to run a failing test with `NVFUSER_DISABLE=parallel_compile NVFUSER_DUMP=segmented_fusion,fusion_ir`. Oftentimes there's some problematic segment. As the segmented fusion is printed in the dependency order, intuitively it's easier to digest what's happening if the segments are compiled in the same order as the dumped output of the segmented fusion. 

Not a strong reason to keep the ordering, but that's the reason I wasn't motivated to change.